### PR TITLE
RELATED: SD-1108 Add new apis to get list of attributeDisplayForms and attributes

### PR DESCRIPTION
**Pull request by dlcd2203**
---

Supported PR commands:

| Command         | Description            |
| --------------- | ---------------------- |
| `ok to test`    | Re-run standard checks |
| `extended test` | BackstopJS tests       |

---

# PR Checklist

-   [x] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [x] `check` passes
-   [x] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by dlcd2203** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/796/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2385/ : SUCCESS in 15m 50s

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/797/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/323/ : UNSTABLE in 20m 04s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/858/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2441/ : SUCCESS in 11m 42s

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 1]
```diff
@@ -220,4 +224,70 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
...
+        const uriIdentifierPairs = await this.authCall((sdk) =>
+            sdk.md.getUrisFromIdentifiers(this.workspace, identifiers),
+        );
+        return uriIdentifierPairs.map(convertUriIdentifierPair);
+    };
+
+    public getAttributeDisplayFormsByUris = async (
+        uris: Uri[],
+    ): Promise<IAttributeDisplayFormMetadataObject[]> => {
+        const wrappedAttributeOrDisplayForms = await this.authCall((sdk) =>
```
Comment (ptquang86) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:238]: so this contains Attribute Or DisplayForms, right?

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 2]
```diff
@@ -220,4 +224,70 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
...
+                    df
+                        .attribute(refDisplayForm)
+                        .title(displayForm.meta.title)
+                        .description(displayForm.meta.summary)
+                        .default(displayForm.content.default)
+                        .id(displayForm.meta.identifier)
+                        .uri(displayForm.meta.uri)
+                        .displayFormType(displayForm.content.type),
+                ),
+            );
```
Comment (ptquang86) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:280]: this and the code in getAttributeDisplayFormsByUris look similar

File: libs/sdk-backend-bear/src/convertors/fromBackend/MetaConverter.ts [hunk 3]
```diff
@@ -122,3 +123,9 @@ export const convertMetadataObjectXrefEntry = (
         deprecated: entry.deprecated === "1",
     };
 };
+
+export const convertUriIdentifierPair = (uriIdentifierPairs: IUriIdentifierPair): IUriIdentifierObject => {
+    return {
+        ...uriIdentifierPairs,
+    };
+};
```
Comment (ptquang86) [inline libs/sdk-backend-bear/src/convertors/fromBackend/MetaConverter.ts:131]: this code look wierd

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 4]
```diff
@@ -220,4 +224,70 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
             }),
         );
     }
+
+    public getUriIdentifierObjects = async (identifiers: Identifier[]): Promise<IUriIdentifierObject[]> => {
```
Comment (ptquang86) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:228]: IMO, use the name `get...By...` is easier to understand

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 5]
```diff
@@ -220,4 +224,70 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
...
+    public getAttributeDisplayFormsByUris = async (
+        uris: Uri[],
+    ): Promise<IAttributeDisplayFormMetadataObject[]> => {
+        const wrappedAttributeOrDisplayForms = await this.authCall((sdk) =>
+            sdk.md.getObjects<GdcMetadata.IWrappedAttributeDisplayForm>(this.workspace, uris),
+        );
+        return wrappedAttributeOrDisplayForms.map((wrapperDisplayForm) => {
+            const displayFormDetails = wrapperDisplayForm.attributeDisplayForm;
+            const ref: UriRef = uriRef(displayFormDetails.meta.uri);
+            const attrRef = uriRef(displayFormDetails.content.formOf);
```
Comment (lvkhanh) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:244]:
```suggestion
    const {
        attributeDisplayForm: {
            meta: { title, summary, identifier, uri },
            content: { default , formOf }
        },
    } = wrapperDisplayForm;
    const ref: UriRef = uriRef(uri);
    const attrRef: UriRef = uriRef(formOf);
```

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 6]
```diff
@@ -220,4 +224,70 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
...
+        const wrappedAttributes = await this.authCall((sdk) =>
+            sdk.md.getObjects<GdcMetadata.IWrappedAttribute>(this.workspace, uris),
+        );
+
+        return wrappedAttributes.map((wrappedAttribute) => {
+            const {
+                meta: { title, uri, isProduction, identifier, summary },
+                content: { displayForms },
+            } = wrappedAttribute.attribute;
+            const refDisplayForm = uriRef(displayForms[0].content.formOf);
```
Comment (lvkhanh) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:268]: `displayForms[0].content.formOf `that's mean attribute `uri` so Should refDisplayForm be first displayFrom uri item or default displayFrom uri item?

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 1 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: fixed

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 2 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: fixed

File: libs/sdk-backend-bear/src/convertors/fromBackend/MetaConverter.ts (hunk 3 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/convertors/fromBackend/MetaConverter.ts:131]: This just converts `IUriIdentifierPair` from `api-client-bear` to `IUriIdentifierObject` from `sdk-backend-spi`

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/879/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2467/ : SUCCESS in 22m 43s

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 6 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: fixed

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 5 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: fixed

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 4 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: changed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/880/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2468/ : SUCCESS in 14m 40s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/882/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2470/ : SUCCESS in 14m 54s

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 7]
```diff
@@ -235,59 +235,64 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
...
-                    .title(displayFormDetails.meta.title)
-                    .description(displayFormDetails.meta.summary)
-                    .default(displayFormDetails.content.default)
-                    .id(displayFormDetails.meta.identifier)
-                    .uri(displayFormDetails.meta.uri),
-            );
-        });
+        return wrappedAttributeDisplayForms.map(
+            (
+                wrapperDisplayForm: GdcMetadata.IWrappedAttributeDisplayForm,
```
Comment (ptquang86) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:243]: how about wrapperDisplayForm  > wrappedDisplayForm?

File: libs/sdk-backend-bear/src/convertors/fromBackend/MetaConverter.ts (hunk 3 above)
Comment (ptquang86) [reply libs/sdk-backend-bear/src/convertors/fromBackend/MetaConverter.ts:131]: Why don't you reuse IUriIdentifierObject?

**Comment by dlcd2203** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/888/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/345/ : SUCCESS in 35m 16s

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts [hunk 8]
```diff
@@ -220,4 +224,75 @@ export class BearWorkspaceMetadata implements IWorkspaceMetadata {
             }),
         );
     }
+
+    public getUrisByIdentifiers = async (identifiers: Identifier[]): Promise<IUriIdentifierObject[]> => {
+        const uriIdentifierPairs = await this.authCall((sdk) =>
+            sdk.md.getUrisFromIdentifiers(this.workspace, identifiers),
+        );
+        return uriIdentifierPairs.map(convertUriIdentifierPair);
+    };
```
Comment (lupko) [inline libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts:233]: we have a function in `utils/api` for this: objRefsToUris. on input the services should be accepting the opaque objRef, then use this function to do the conversion.

File: libs/sdk-backend-spi/src/workspace/metadata/index.ts [hunk 9]
```diff
@@ -58,4 +61,28 @@ export interface IWorkspaceMetadata {
...
+     */
+    getAttributesByUris(uris: Uri[]): Promise<IAttributeMetadataObject[]>;
+
+    /**
+     * Gets the list of metadata objects that contains only uri and identifier.
+     *
+     * @param identifiers -  list of identifiers of attribute.
+     * @returns promise of list of object.
+     */
+    getUrisByIdentifiers(identifiers: Identifier[]): Promise<IUriIdentifierObject[]>;
```
Comment (lupko) [inline libs/sdk-backend-spi/src/workspace/metadata/index.ts:87]: functions like this are not acceptable on the SPI level. instead, you should use the opaque ObjRef (which can be either uri or id) and then have the backend handle them and do what is needed.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/913/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2498/ : SUCCESS in 11m 36s

**Comment by dlcd2203** [comment]
extended test

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/914/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/351/ : SUCCESS in 25m 46s

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 8 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: I have removed the function.

File: libs/sdk-backend-spi/src/workspace/metadata/index.ts (hunk 9 above)
Comment (dlcd2203) [reply libs/sdk-backend-spi/src/workspace/metadata/index.ts]: I have updated the PR.

**Comment by dlcd2203** [comment]
Hi @lupko . I have updated my PR. Please help me review it again. Thanks

File: libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts (hunk 7 above)
Comment (dlcd2203) [reply libs/sdk-backend-bear/src/backend/workspace/metadata/index.ts]: fixed

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/916/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2500/ : SUCCESS in 11m 51s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Review by kandl — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/362/ : SUCCESS in 30s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/618/ : SUCCESS in 15s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/2512/ : SUCCESS in 12m 06s

**Pull request merged**